akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala

```diff
@@ -753,6 +768,158 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
+        //... and it is possible to consume all responses (from stream 1 to stream 7)
+        val resp3: HttpResponse = user.expectResponse()
+        val resp5: HttpResponse = user.expectResponse()
+        val resp7: HttpResponse = user.expectResponse()
+
+        private val eventualEntities: Seq[Future[HttpEntity.Strict]] =
+          Seq(resp1, resp3, resp5, resp7).map(_.entity.toStrict(100.millis))
+        private val entities: Seq[HttpEntity.Strict] = Future.sequence(eventualEntities).futureValue
+        // The complete entity of the responses is available
+        entities shouldBe Seq(
+          HttpEntity.Empty, // 1 has no data
+          HttpEntity.Empty, // 3 has no data
+          HttpEntity.Strict(ContentTypes.`application/octet-stream`, ByteString("Hello World 5!")),
+          HttpEntity.Strict(ContentTypes.`application/octet-stream`, ByteString("Hello World 7!"))
+        )
+
+        // Completion doesn't happen until entities are consumed
+        user.responseIn.expectComplete()
+
+      }
+
+      "until in-flight streaming requests are fully sent (after response is already complete) (uncommon)" inAssertAllStagesStopped new TestSetup(
+        Setting(SettingIdentifier.SETTINGS_MAX_CONCURRENT_STREAMS, 3)
+      ) with NetProbes {
+        val requestStreamProbe = TestPublisher.probe[ByteString]()
+        val requestEntity = HttpEntity(ContentTypes.`application/octet-stream`, Source.fromPublisher(requestStreamProbe))
+
+        // Given several in-flight requests...
+        val request = HttpRequest(method = HttpMethods.POST, uri = "https://www.example.com/", entity = requestEntity)
+        user.emitRequest(request)
+
+        network.expectDecodedResponseHEADERSPairs(1, endStream = false)
+
+        // ... and some request already got their response (canary round-trip)
+        network.sendFrame(HeadersFrame(streamId = 1, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None))
+        user.expectResponse()
+
+        requestStreamProbe.sendNext(ByteString("hello"))
+        network.expectDATA(1, endStream = false, ByteString("hello"))
+
+        // now closing user out
+        user.requestOut.sendComplete()
+
+        // stage needs to stay alive to send rest of request entity
+        requestStreamProbe.sendNext(ByteString("world"))
+        network.expectDATA(1, endStream = false, ByteString("world"))
+
+        // queueing more data in the multiplexer without immediately fetching it on the network
+        requestStreamProbe.sendNext(ByteString("almost"))
+        requestStreamProbe.sendNext(ByteString(" done"))
+        requestStreamProbe.sendComplete()
+        network.expectDATA(1, endStream = true, ByteString("almost done"))
+
+        // Completion doesn't happen until entities are sent
+        user.responseIn.expectComplete()
+        network.toNet.expectComplete()
+        network.fromNet.expectCancellation()
+      }
```

Comment:
This is a nice test. I didn't think of the scenario and making the client await until both outgoing and incoming streams complete is the most reasonable behavior.